FlowCrypt/src/main/java/com/flowcrypt/email/core/msg/RawBlockParser.kt

```diff
@@ -0,0 +1,309 @@
+          RawBlockType.HTML_TEXT,
+          part.inputStream.readBytes(),
+          isOpenPGPMimeSigned
+        )
+      )
+      else -> detectBlocks(part.inputStream.readText()).map {
+        it.copy(isOpenPGPMimeSigned = isOpenPGPMimeSigned)
+      }
+    }
+  }
+
+  fun detectBlocks(source: String): Collection<RawBlock> {
+    val normalized = source.normalize()
+    val blocks = mutableListOf<RawBlock>()
+    var startAt = 0
+    while (true) {
+      val continueAt = detectNextBlock(normalized, startAt, blocks)
+      if (startAt >= continueAt) return blocks
+      startAt = continueAt
+    }
+  }
+
+  private fun treatAs(mimePart: MimePart): TreatAs {
+    val name = mimePart.fileName?.lowercase() ?: ""
+    val baseContentType = mimePart.baseContentType()
+    val length = mimePart.size
+    when {
+      name in HIDDEN_FILE_NAMES -> {
+        // PGPexch.htm.pgp is html alternative of textual body content produced
+        // by the PGP Desktop and GPG4o
+        return TreatAs.HIDDEN
+      }
+
+      "" == name && baseContentType?.startsWith("image/") != true -> {
+        return if (length < 100) TreatAs.SIGNATURE else TreatAs.PGP_MSG
+      }
+
+      "signature.asc" == name || "application/pgp-signature" == baseContentType -> {
+        return TreatAs.SIGNATURE
+      }
+
+      "msg.asc" == name && length < 100 && "application/pgp-encrypted" == baseContentType -> {
+        // mail.ch does this - although it looks like encrypted msg,
+        // it will just contain PGP version eg "Version: 1"
+        return TreatAs.SIGNATURE
+      }
+
+      "Message.pgp" == mimePart.fileName || name in ENCRYPTED_MSG_NAMES -> {
+        return TreatAs.PGP_MSG
+      }
+
+      ENCRYPTED_FILE_REGEX.containsMatchIn(name) -> {
+        // ends with one of .gpg, .pgp, .???.asc, .????.asc
+        return TreatAs.ENCRYPTED_FILE
+      }
+
+      PRIVATE_KEY_REGEX.containsMatchIn(name) -> {
+        return TreatAs.PRIVATE_KEY
+      }
+
+      "application/pgp-keys" == baseContentType -> {
+        return TreatAs.PUBLIC_KEY
+      }
+
+      PUBLIC_KEY_REGEX_1.containsMatchIn(name) -> {
+        // name starts with a key id
+        return TreatAs.PUBLIC_KEY
+      }
+
+      name.contains("public") && PUBLIC_KEY_REGEX_2.containsMatchIn(name) -> {
+        // name contains the word "public", any key id and ends with .asc
+        return TreatAs.PUBLIC_KEY
+      }
+
+      name.endsWith(".asc") && length > 0 && checkForPublicKeyBlock(mimePart.inputStream) -> {
+        return TreatAs.PUBLIC_KEY
+      }
+
+      name.endsWith(".asc") && length < 100000 && !mimePart.isInline() -> {
+        return TreatAs.PGP_MSG
+      }
+
+      else -> {
+        return TreatAs.PLAIN_FILE
+      }
+    }
+  }
+
+  private fun checkForPublicKeyBlock(stream: InputStream): Boolean {
+    val a = ByteArray(100)
+    val r = stream.read(a)
+    if (r < 1) return false
+    val s = String(if (r == a.size) a else a.copyOf(r), StandardCharsets.US_ASCII)
```

Comment:
a.size <= r

Reply:
Are you sure? It's cut-paste from PgpMsg. I didn't change this code, just copied it.

Reply:
Yes, I understand, but the above is a bit more optimal than what it is now.